chalk-solve/src/infer/unify.rs

```diff
@@ -57,86 +62,129 @@ impl<'t, I: Interner> Unifier<'t, I> {
+                        Ok(())
+                    }
+                    Variance::Contravariant => {
+                        self.push_subtype_goal(b.clone(), a.clone());
+                        Ok(())
+                    }
                 }
             }
 
             // Unifying `forall<X> { T }` with some other forall type `forall<X> { U }`
             (&TyKind::Function(ref fn1), &TyKind::Function(ref fn2)) => {
                 if fn1.sig == fn2.sig {
-                    self.unify_binders(fn1, fn2)
+                    Zip::zip_with(
+                        self,
+                        variance,
+                        &fn1.clone().into_binders(interner),
+                        &fn2.clone().into_binders(interner),
+                    )
                 } else {
                     Err(NoSolution)
                 }
             }
 
             (&TyKind::Placeholder(ref p1), &TyKind::Placeholder(ref p2)) => {
-                Zip::zip_with(self, p1, p2)
+                Zip::zip_with(self, variance, p1, p2)
             }
 
             // Unifying two dyn is possible if they have the same bounds.
-            (&TyKind::Dyn(ref qwc1), &TyKind::Dyn(ref qwc2)) => Zip::zip_with(self, qwc1, qwc2),
+            (&TyKind::Dyn(ref qwc1), &TyKind::Dyn(ref qwc2)) => {
+                Zip::zip_with(self, variance, qwc1, qwc2)
+            }
 
             (TyKind::BoundVar(_), _) | (_, TyKind::BoundVar(_)) => panic!(
                 "unification encountered bound variable: a={:?} b={:?}",
                 a, b
             ),
 
             // Unifying an alias type with some other type `U`.
-            (_, &TyKind::Alias(ref alias)) => self.unify_alias_ty(alias, a),
-            (&TyKind::Alias(ref alias), _) => self.unify_alias_ty(alias, b),
+            (_, &TyKind::Alias(ref alias)) => self.relate_alias_ty(variance.invert(), alias, a),
+            (&TyKind::Alias(ref alias), _) => self.relate_alias_ty(variance, alias, b),
+
+            // FIXME: needs to handle relating a var and ty; needs generalization
+            // Relating an inference variable with a non-inference variable.
+            (a_data @ &TyKind::InferenceVar(_, _), b_data @ _)
+            | (a_data @ _, b_data @ &TyKind::InferenceVar(_, _)) => {
+                // Correct variance when we flip a/b.
+                let (new_variance, var, kind, ty_data) = match (a_data, b_data) {
+                    (&TyKind::InferenceVar(var, kind), ty_data @ _) => {
+                        (variance, var, kind, ty_data)
+                    }
+                    (ty_data @ _, &TyKind::InferenceVar(var, kind)) => {
+                        // var is grabbed from 'b', but given in as the left
+                        // parameter to relate_var_ty, so we need to flip variance.
+                        (variance.invert(), var, kind, ty_data)
+                    }
+                    _ => unreachable!(),
+                };
```

Comment:
It would probably be cleaner to just have two different match arms here...

The code below can probably just be moved into `relate_var_ty` and make it return a `Result` so we can move the interner/float check into it